hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/validate/HazelcastSqlValidator.java

```diff
@@ -525,4 +545,100 @@ public CalciteContextException newValidationError(SqlNode node, Resources.ExInst
         }
         return exception;
     }
+
+    /**
+     * Wraps TABLE operators in subqueries when they appear as join operands.
+     * <p>
+     * {@code FROM TABLE(...) JOIN TABLE(...)} → <br>
+     * {@code FROM (SELECT * FROM TABLE(...)) JOIN (SELECT * FROM TABLE(...))}
+     */
+    private static class TableOperatorWrapper extends SqlIdentityVisitor {
+        @Override
+        public SqlNode visit(@Nonnull SqlCall call) {
+            if (call instanceof SqlJoin) {
+                SqlJoin join = (SqlJoin) call;
+                join.setLeft(wrapTableOperator(join.getLeft()));
+                join.setRight(wrapTableOperator(join.getRight()));
+                return join;
+            }
+            return super.visit(call);
+        }
+
+        private SqlNode wrapTableOperator(SqlNode node) {
+            if (node instanceof SqlCall) {
+                SqlCall call = (SqlCall) node;
+                if (call.getOperator().getKind() == AS) {
+                    call.setOperand(0, wrapTableOperator(call.getOperandList().get(0)));
+                    return call;
+                } else if (call.getOperator().getKind() == COLLECTION_TABLE) {
+                    return new SqlSelect(call.getParserPosition(), null, SqlNodeList.SINGLETON_STAR,
+                            super.visit(call), null, null, null, null, null, null, null, null);
+                }
+            }
+            return node.accept(this);
+        }
+    }
+
+    /**
+     * An {@link SqlVisitor} that replaces every {@link SqlNode} by itself,
+     * which makes it suitable for implementing {@code SqlNode} transformers.
+     * Note that such a transformer needs to be called in
+     * {@link #performUnconditionalRewrites}.
+     */
+    private static class SqlIdentityVisitor implements SqlVisitor<SqlNode> {
```

Comment:
Later, I realized that there is `SqlShuttle` for this purpose, but `SqlIdentityVisitor` is expected to be slightly more performant since it doesn't recreate calls/lists on update and performs in-place edit. However, for the sake of maintainability, e.g. in order not to track the buggy behavior of `SqlSelect.setOperand`, we can just rely on `SqlShuttle`.